cloud-controller-manager/do/loadbalancers.go

```diff
@@ -866,12 +866,32 @@ func buildForwardingRule(service *v1.Service, port *v1.ServicePort, protocol, ce
 }
 
 func buildFirewall(service *v1.Service) *godo.LBFirewall {
+	var allow []string
+	sourceRanges := getSourceRangeRules(service)
+	if len(sourceRanges) > 0 {
+		// source ranges should take precedence over annotation based allow rules
+		allow = sourceRanges
+	} else {
+		allow = getStrings(service, annDOAllowRules)
+	}
 	return &godo.LBFirewall{
 		Deny:  getStrings(service, annDODenyRules),
-		Allow: getStrings(service, annDOAllowRules),
+		Allow: allow,
 	}
 }
 
+// getSourceRangeRules returns loadBalancerSourceRanges in the format
+// `cidr:{source}`
+func getSourceRangeRules(service *v1.Service) []string {
+	sourceRanges := service.Spec.LoadBalancerSourceRanges
+
+	for i, sourceRange := range sourceRanges {
+		sourceRanges[i] = "cidr:" + sourceRange
```

Comment:
Curious, is the source range format standardized / specified anyhow? The [API documentation](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceSpec) seems scarce on this point. IIUC, we're assuming `<v.w.x.y/z>`?

Should we validate the entries here so that we don't need an extra roundtrip to the LB API to validate?

Reply:
I haven't found anything that would document a certain format. However if you follow the code from [this](https://github.com/kubernetes/kubernetes/blob/8565e375251450a291f0af3c6195c7a5bf890292/pkg/api/service/util.go#L44) then you'll see that the parsing function only allows the '/' CIDR notation, which is what I would expect give the field name.
We could maybe think about re-using the function which would make sure it parses like upstream wants it to?
There's the same function in the [cloud-provider package](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/4e844752045f0fbe4925723142c8ae6bfa949d92/vendor/k8s.io/cloud-provider/service/helpers/helper.go#L60-L60) which is already imported.

Reply:
Yes, lets do that. I also don't want to adopt yet another annotation.